File: buildSrc/src/main/java/io/github/stackunderflow111/jooqflywaytestcontainersdemo/database/buildservices/ExternalDatabase.java
package io.github.stackunderflow111.jooqflywaytestcontainersdemo.database.buildservices;

import io.github.stackunderflow111.jooqflywaytestcontainersdemo.database.Database;
import org.gradle.api.provider.Property;
import org.gradle.api.services.BuildServiceParameters;

abstract public class ExternalDatabase implements Database<ExternalDatabase.Params> {

    public interface Params extends BuildServiceParameters {
        Property<String> getJdbcUrl();

        Property<String> getUsername();

        Property<String> getPassword();
    }

    @Override
    public String getJdbcUrl() {
        return getParameters().getJdbcUrl().get();
    }

    @Override
    public String getUsername() {
        return getParameters().getUsername().get();
    }

    @Override
    public String getPassword() {
        return getParameters().getPassword().get();
    }

}

File: buildSrc/src/main/kotlin/io/github/stackunderflow111/jooqflywaytestcontainersdemo/buildservices/ExternalDatabase.kt
package io.github.stackunderflow111.jooqflywaytestcontainersdemo.buildservices

import org.gradle.api.provider.Property
import org.gradle.api.services.BuildServiceParameters

abstract class ExternalDatabase : Database<ExternalDatabase.Params> {
  interface Params : BuildServiceParameters {

    fun getJdbcUrl(): Property<String>
    fun getUsername(): Property<String>
    fun getPassword(): Property<String>
  }

  override val jdbcUrl: String
    get() = parameters.getJdbcUrl().get()
  override val username: String
    get() = parameters.getUsername().get()
  override val password: String
    get() = parameters.getPassword().get()
}